cpp/cuproj/tests/wgs_to_utm_test.cu

```diff
@@ -0,0 +1,221 @@
+
+  auto run = [&]() {
+    expected_coords = input_coords;
+    run_proj_test(expected_coords, epsg_src, epsg_dst);
+
+    auto proj = cuproj::make_projection<coordinate<T>>(epsg_src, epsg_dst);
+
+    run_cuproj_test(input_coords, expected_coords, proj, cuproj::direction::FORWARD, tolerance);
+    run_cuproj_test(expected_coords, input_coords, proj, cuproj::direction::INVERSE, tolerance);
+  };
+
+  // forward construction
+  run();
+  // invert construction
+  input_coords = expected_coords;
+  std::swap(epsg_src, epsg_dst);
+  run();
+}
+
+// Just test construction of the projection from supported EPSG codes
+TYPED_TEST(ProjectionTest, make_projection_valid_epsg)
+{
+  using T = TypeParam;
+  cuproj::make_projection<coordinate<T>>("EPSG:4326", "EPSG:32756");
+  cuproj::make_projection<coordinate<T>>("EPSG:32756", "EPSG:4326");
+  cuproj::make_projection<coordinate<T>>("EPSG:4326", "EPSG:32601");
+  cuproj::make_projection<coordinate<T>>("EPSG:32601", "EPSG:4326");
+}
+
+// Test that construction of the projection from unsupported EPSG codes throws
+// expected exceptions
+TYPED_TEST(ProjectionTest, invalid_epsg)
+{
+  using T = TypeParam;
+  EXPECT_THROW(cuproj::make_projection<coordinate<T>>("EPSG:4326", "EPSG:756"),
+               cuproj::logic_error);
+  EXPECT_THROW(cuproj::make_projection<coordinate<T>>("EPSG:4326", "UTM:32756"),
+               cuproj::logic_error);
+}
+
+// Test on a single coordinate
+TYPED_TEST(ProjectionTest, one)
+{
+  using T = TypeParam;
+
+  coordinate<T> sydney{-33.865143, 151.209900};  // Sydney, NSW, Australia
+  std::vector<coordinate<T>> input{sydney};
+  // We can expect nanometer accuracy with double precision. The precision ratio of
+  // double to single precision is 2^53 / 2^24 == 2^29 ~= 10^9, then we should
+  // expect meter (10^9 nanometer) accuracy with single precision.
+  T tolerance = std::is_same_v<T, float> ? T{1.0} : T{1e-9};
+  run_forward_and_inverse<T>(input, tolerance);
+}
+
+// Generate a grid of coordinates
+template <typename T>
+struct grid_generator {
+  coordinate<T> min_corner{};
+  coordinate<T> max_corner{};
+  coordinate<T> spacing{};
+  int num_points_x{};
+  int num_points_y{};
+
+  grid_generator(coordinate<T> min_corner,
+                 coordinate<T> max_corner,
+                 int num_points_x,
+                 int num_points_y)
+    : min_corner(min_corner),
+      max_corner(max_corner),
+      num_points_x(num_points_x),
+      num_points_y(num_points_y)
+  {
+    spacing = coordinate<T>{(max_corner.x - min_corner.x) / num_points_x,
+                            (max_corner.y - min_corner.y) / num_points_y};
+  }
+
+  __device__ coordinate<T> operator()(int i) const
+  {
+    return min_corner +
+           coordinate<T>{(i % num_points_x) * spacing.x, (i / num_points_x) * spacing.y};
+  }
+};
+
+// Test on a grid of coordinates
+TYPED_TEST(ProjectionTest, many)
+{
+  using T = TypeParam;
+  // generate a grid of (lat, lon) coordinates around Sydney Harbour
```

Comment:
Why not test on a world wide grid?

Reply:
Because UTM zones are 6 degrees wide and accuracy would suffer outside that.

Reply:
I've added test grids around multiple positions on the globe for more extensive testing.